### PR TITLE
Guard against FPEs in the PTE solver

### PR DESCRIPTION
**Pull request by jhp-lanl**
## PR Summary

This PR mostly adds `robust::ratio` divides throughout the PTE solver. This was especially important for two scenarios:
1. When the `RhoPmin()` function of an EOS returns zero (which is the default)
2. When the specific internal energy is zero (rare, but not disallowed)

In connection with the internal energy, I also made sure that the energy scale, `uscale` is non-negative since it didn't seem to make sense to me for this to be a signed quantity. _**Please**_ let me know if I'm wrong here, but it always appeared to be a positive scaling factor since it appeared that important signs would cancel in some circumstances (negative pressures for example).

Some debug asserts were also added to codify the necessary conditions for the PTE solver to operate. For example, one could reasonably assume that the PTE solver is robust to zero volume or mass fractions, but that isn't actually the case.

Finally, with the huge help of @rbberger , I corrected a CMake bug where a statically-built Kokkos wouldn't properly link to the Fortran tests.

Fixes https://github.com/lanl/singularity-eos/issues/354

## PR Checklist

- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- N/A Document any new features, update documentation for changes made.
- [x] Make sure the copyright notice on any files you modified is up to date.
- [x] After creating a pull request, note it in the CHANGELOG.md file.
- [ ] LANL employees: make sure tests pass both on the github CI and on the Darwin CI

If preparing for a new release, in addition please check the following:
- [ ] Update the version in cmake.
- [ ] Move the changes in the CHANGELOG.md file under a new header for the new release, and reset the categories.

### CODE REVIEW

**Comment by jhp-lanl** [comment]
All 54 tests pass in my local testing

File: CMakeLists.txt
Comment (Yurlungur) [inline CMakeLists.txt:338]: Maybe worth commenting that this is to make cray happy.

File: CMakeLists.txt
Comment (Yurlungur) [inline CMakeLists.txt:405]: I'm very confused by why this works but... :shrug:

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (Yurlungur) [inline singularity-eos/closure/mixed_cell_models.hpp:175]: :+1: for debug only guards on sanity.

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (Yurlungur) [inline singularity-eos/closure/mixed_cell_models.hpp:260]: :+1:

**Review by Yurlungur — APPROVED** [review]
Thanks for going through this @jhp-lanl ! This is a much-needed cleanup.

**Comment by Yurlungur** [comment]
> All 54 tests pass in my local testing

Does re-git pass? If so, I think this is good to go.

File: CMakeLists.txt
Comment (jhp-lanl) [reply CMakeLists.txt:338]: I'm not 100% that's true to be honest. I'll try the same process on a non-cray machine and see if I get the same result.

We weren't able to reproduce on non-cray machines but that wasn't with the EAP build process.

**Comment by jhp-lanl** [comment]
> > All 54 tests pass in my local testing
> 
> Does re-git pass? If so, I think this is good to go.

I think I'll need some CI changes from @rbberger before it will pass. @rbberger can I merge in the `cmake_fixup_static_linking` branch or are you planning to put that into main first?

File: CMakeLists.txt
Comment (jhp-lanl) [reply CMakeLists.txt:405]: Me too, but I seem to recall linking issues with the wrong scope before. I _think_ this just gets the libraries later in the link command, but I'm not 100% sure. I defer to @rbberger

Comment (rbberger) [reply CMakeLists.txt:405]: You can think of this as telling CMake to add an explicit dependency to Kokkos (via Common) when compiling the library itself and to anything that links to it. It's a missing edge in the dependency graph that lets CMake determine the proper link line.

File: CMakeLists.txt
Comment (jhp-lanl) [reply CMakeLists.txt:338]: Yeah, I'm going to push back on this comment. The Fortran interface test _is_ dependent on Kokkos when we build with Kokkos, and I don't think that's covered by the `*Interface` library. I think that's why it needs to be linked to the `*Common` library.

**Comment by rbberger** [comment]
> I think I'll need some CI changes from @rbberger before it will pass. @rbberger can I merge in the `cmake_fixup_static_linking` branch or are you planning to put that into main first?
@ktsai7 and I are working on the CI. I would say if this works for you, merge it into `main`. I'll create my own Github PR for the remaining things that touch the CI and Spackage.

File: CMakeLists.txt
Comment (Yurlungur) [reply CMakeLists.txt:338]: Ah ok. Maybe add a comment to that effect? I'm mostly concerned about someone changing it back, since this change is non-obvious.

**Comment by Yurlungur** [comment]
> > I think I'll need some CI changes from @rbberger before it will pass. @rbberger can I merge in the `cmake_fixup_static_linking` branch or are you planning to put that into main first?
> 
> @ktsai7 and I are working on the CI. I would say if this works for you, merge it into `main`. I'll create my own Github PR for the remaining things that touch the CI and Spackage.

Sounds good. In that case, merge when ready @jhp-lanl

File: CHANGELOG.md
Comment (jhp-lanl) [inline CHANGELOG.md:19]: FYI - this was just a duplicated section for some reason. I combined above

File: CMakeLists.txt
Comment (Yurlungur) [reply CMakeLists.txt:340]: LGTM

**Comment by jhp-lanl** [comment]
> @jhp-lanl what's the status on this? Is this ready for merge?

Ah yes, I was waiting for Richard's CI merge and then dropped the ball. Let me resolve the conflicts and get it merged ASAP

**Pull request merged**
